util/slices.go

```diff
@@ -1,35 +1,19 @@
 package util
 
-import (
-	"golang.org/x/exp/constraints"
-)
-
 // CopySlice copies the slice to a newly-allocated one
 func CopySlice[T any](slice []T) []T { // xx:inline
 	return append([]T(nil), slice...)
 }
 
-// EachInSlice calls the given func for each of (index, value) pair in the given slice
-func EachInSlice[T any](slice []T, action func(index int, item T)) {
-	for index, item := range slice {
-		action(index, item)
-	}
-}
-
-// MapSlice transforms the given slice by mapping each item to something else
-func MapSlice[T any, R any](slice []T, mapper func(item T) R) []R {
-	output := make([]R, len(slice))
+// MapSliceTo2 transforms the given slice by mapping each item to two different items.
+//
+// This is needed because multi-value result cannot be reprersented by a type itself in Go,
+// making it impossible to chain deconstruction steps separately.
+func MapSliceTo2[T any, R1 any, R2 any](slice []T, mapper func(item T) (R1, R2)) ([]R1, []R2) {
```

Comment:
Why can't we have our data structured beforehand? Why do we have to convert 1 slice into 2 slices?

It looks like we are doing a double work and making code more complex:
- we are building data in preparePipeline function already
- we are creating outputNames [here](https://github.com/relex/slog-agent/blob/2cd08174fa35de6f5b7d2df91580498a0aaa661c/test/pipeline.go#L47) which involves a loop already
- then we are doing another round of iterations with this MapSliceTo2 function [here](https://github.com/relex/slog-agent/blob/2cd08174fa35de6f5b7d2df91580498a0aaa661c/test/integration_test.go#L58)

Is it possible to return this 2 slices from preparePipeline function and avoid double loops??

Reply:
Yes, because originally `preparePipeline` doesn't deal with output overrides / chunk saver (was optional). But now output overrides are always used with the pipeline, since it's only used for benchmarking and testing

It can be structured and simplified yes, need to think a bit.